test/parallel-benchmark/mzcompose.py

```diff
@@ -0,0 +1,1190 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+"""
+Benchmark with scenarios combining closed and open loops, can run multiple
+actions concurrently, measures various kinds of statistics.
+"""
+
+import collections
+import gc
+import queue
+import random
+import ssl
+import threading
+import time
+from collections.abc import Iterator
+from dataclasses import dataclass
+from pathlib import Path
+from textwrap import dedent
+from urllib.parse import parse_qs, unquote, urlparse
+
+import matplotlib.pyplot as plt
+import numpy
+import pg8000
+from matplotlib.markers import MarkerStyle
+
+from materialize import MZ_ROOT, buildkite
+from materialize.mzcompose.composition import Composition, WorkflowArgumentParser
+from materialize.mzcompose.services.balancerd import Balancerd
+from materialize.mzcompose.services.clusterd import Clusterd
+from materialize.mzcompose.services.cockroach import Cockroach
+from materialize.mzcompose.services.kafka import Kafka as KafkaService
+from materialize.mzcompose.services.kgen import Kgen as KgenService
+from materialize.mzcompose.services.materialized import Materialized
+from materialize.mzcompose.services.minio import Minio
+from materialize.mzcompose.services.mysql import MySql
+from materialize.mzcompose.services.postgres import Postgres
+from materialize.mzcompose.services.redpanda import Redpanda
+from materialize.mzcompose.services.schema_registry import SchemaRegistry
+from materialize.mzcompose.services.testdrive import Testdrive
+from materialize.mzcompose.services.zookeeper import Zookeeper
+from materialize.mzcompose.test_result import (
+    FailedTestExecutionError,
+    TestFailureDetails,
+)
+from materialize.util import all_subclasses
+
+SERVICES = [
+    Zookeeper(),
+    KafkaService(),
+    SchemaRegistry(),
+    Redpanda(),
+    Cockroach(setup_materialize=True),
+    Minio(setup_materialize=True),
+    KgenService(),
+    Postgres(),
+    MySql(),
+    Balancerd(),
+    # Overridden below
+    Materialized(),
+    Clusterd(),
+    Testdrive(no_reset=True, seed=1),
+]
+
+
+@dataclass
+class PgConnInfo:
+    user: str
+    host: str
+    port: int
+    database: str
+    password: str | None = None
+    ssl: bool = False
+
+    def connect(self) -> pg8000.Connection:
+        return pg8000.connect(
+            host=self.host,
+            port=self.port,
+            user=self.user,
+            password=self.password,
+            database=self.database,
+            ssl_context=ssl.SSLContext() if self.ssl else None,
+        )
```

Comment:
Shall we move this into some utils file? I guess we have something similar in several places.

Reply:
For example, `MzDbConfig` is also partially similar.